activerecord/lib/active_record/relation.rb

```diff
@@ -304,84 +304,151 @@ def many?
+    # has a limit or offset and you're using MySQL or a version of PostgreSQL or SQLite that
+    # supports window functions), and the timestamp of the last updated record. When a
+    # new record comes to match the query, or any of the existing records is updated or
+    # deleted, the cache version changes.
     #
     # If the collection is loaded, the method will iterate through the records
     # to generate the timestamp, otherwise it will trigger one SQL query like:
     #
     #    SELECT COUNT(*), MAX("products"."updated_at") FROM "products" WHERE (name like '%Cosmic Encounter%')
-    def cache_version(timestamp_column = :updated_at)
+    def cache_version(deprecated_timestamp_column = nil, id_column: primary_key, timestamp_column: :updated_at)
+      if deprecated_timestamp_column
+        ActiveSupport::Deprecation.warn("Passing a timestamp column as a positional argument to ActiveRecord::Relation#cache_version is deprecated. Please use a `:timestamp_column` keyword argument instead.")
+        timestamp_column = deprecated_timestamp_column
+      end
+
       if collection_cache_versioning
         @cache_versions ||= {}
-        @cache_versions[timestamp_column] ||= compute_cache_version(timestamp_column)
+        @cache_versions[[id_column, timestamp_column]] ||= compute_cache_version(id_column: id_column, timestamp_column: timestamp_column)
       end
     end
 
-    def compute_cache_version(timestamp_column) # :nodoc:
+    def compute_cache_version(id_column: primary_key, timestamp_column: :updated_at) # :nodoc:
+      size = 0
+      first_id = nil
+      last_id = nil
+      timestamp = nil
+
       if loaded? || distinct_value
         size = records.size
+
         if size > 0
-          timestamp = max_by(&timestamp_column)._read_attribute(timestamp_column)
+          timestamp = max_by(&timestamp_column)._read_attribute(timestamp_column).utc.to_s(cache_timestamp_format)
+
+          if has_limit_or_offset?
+            first_id = first._read_attribute(id_column)
+            last_id = last._read_attribute(id_column)
+          end
         end
       else
         collection = eager_loading? ? apply_join_dependency : self
-
-        column = connection.visitor.compile(arel_attribute(timestamp_column))
-        select_values = "COUNT(*) AS #{connection.quote_column_name("size")}, MAX(%s) AS timestamp"
+        timestamp_col = connection.visitor.compile(arel_attribute(timestamp_column))
+        id_col = connection.visitor.compile(arel_attribute(id_column))
 
         if collection.has_limit_or_offset?
-          query = collection.select("#{column} AS collection_cache_key_timestamp")
-          subquery_alias = "subquery_for_cache_key"
-          subquery_column = "#{subquery_alias}.collection_cache_key_timestamp"
-          arel = query.build_subquery(subquery_alias, select_values % subquery_column)
+          case connection.adapter_name
+          when "PostgreSQL", "SQLite"
```

Comment:
I'm not familiar with Rails DB-related internals, but I think this kind of branching will need to be abstracted such that every connection adapter can provide its own implementation.  #37944 or #37973 might be useful once merged?

Reply:
Avoiding this type of branching would be ideal, but I feel the code is so specific to this one use case that abstracting it out to the connection adapter level would make the code harder to follow. Leaving it here for now, but welcome pushback.

I am monitoring those pull requests - once merged, we'll have the opportunity to tidy up some of the Arel.

Reply:
After thinking about this some more, moving the Postgres/MySQL/SQLite-specific logic to the adapter level is starting to feel like the correct "Rails" way to structure this code.

I've drafted a refactored PR, accordingly: #38171

I'm still not fully certain which direction to go in, so I'm leaving this PR as is (for now). But I'm prepared to move forward with either option.

Reply:
After chewing on this a little more, I've rebased this PR to move the Postgres/MySQL/SQLite-specific logic into connection adapters.

The previous `ActiveRecord::Relation#compute_cache_version` method is archived here, for reference: https://gist.github.com/alipman88/02f6dd8af2879b597bb9862b9ddc96b8